### PR TITLE
[SYCL][ESIMD] Add test for kernel defined in template function

### PR DESCRIPTION
**Pull request by sarnex**
Related to https://github.com/intel/llvm/pull/7996

Signed-off-by: Sarnie, Nick <nick.sarnie@intel.com>

### CODE REVIEW

**Review by v-klochkov — COMMENTED** [review]
Thank you for the test.
I have only 2 comments:
- PR needs clang-format fix

- Not a requirement to change now, but a notice worse mentioning:
   If you have equal choice between using USM or BUFFERs, then please choose USM because GPU RT is going to drop support of stateful accesses soon; buffers/acc API should still be supported, but via some emulation method (acc.get_pointer() + stateless GenX intrin).

**Comment by sarnex** [comment]
@v-klochkov Thanks!

The clang-format looks wrong acutally, it's breaking a comment:

```
diff --git a/SYCL/ESIMD/template.cpp b/SYCL/ESIMD/template.cpp
index 2a20ebe4..24ab5e98 100644
--- a/SYCL/ESIMD/template.cpp
+++ b/SYCL/ESIMD/template.cpp
@@ -1,4 +1,5 @@
-//==---------------- template.cpp  - DPC++ ESIMD on-device test -------------==//
+//==---------------- template.cpp  - DPC++ ESIMD on-device test
+//-------------==//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
```

I just modified vadd_1d.cpp instead of writing this myself, so I'll just leave it for now

**Comment by v-klochkov** [comment]
> d_1d.cpp instead of writing this myself, so I'll just l

I think making the line 1 symbol shorter would solve the problem with clang-format

**Comment by sarnex** [comment]
> I think making the line 1 symbol shorter would solve the problem with clang-format

ah ok let me try thanks

**Comment by v-klochkov** [comment]
> @v-klochkov Thanks!
> 
> The clang-format looks wrong acutally, it's breaking a comment:
> 
> ```
> diff --git a/SYCL/ESIMD/template.cpp b/SYCL/ESIMD/template.cpp
> index 2a20ebe4..24ab5e98 100644
> --- a/SYCL/ESIMD/template.cpp
> +++ b/SYCL/ESIMD/template.cpp
> @@ -1,4 +1,5 @@
> -//==---------------- template.cpp  - DPC++ ESIMD on-device test -------------==//
> +//==---------------- template.cpp  - DPC++ ESIMD on-device test
> +//-------------==//
>  //
>  // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
>  // See https://llvm.org/LICENSE.txt for license information.
> ```
> 
> I just modified vadd_1d.cpp instead of writing this myself, so I'll just leave it for now

Please keep this test as a separate one. vadd_1d.cpp is known as one of most trivial tests that always worked and perhaps will.

File: SYCL/ESIMD/template.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
+//==---------------- template.cpp  - DPC++ ESIMD on-device test -------------==//
```
Comment (v-klochkov) [inline SYCL/ESIMD/template.cpp]:
```suggestion
//==---------------- template.cpp  - DPC++ ESIMD on-device test ------------==//
```

**Comment by sarnex** [comment]
> Please keep this test as a separate one. vadd_1d.cpp is known as one of most trivial tests that always worked and perhaps will.

Sorry, I mean this test is based on that one. That file is not modified.

**Comment by v-klochkov** [comment]
@sarnex - please click 'Ready for review' when you think it is ready.

**Comment by sarnex** [comment]
@v-klochkov Done, whoops

**Review by v-klochkov — APPROVED** [review]

**Comment by v-klochkov** [comment]
The new test expectedly failed without the fix in ESIMD headers and passed with it (Jenkins/llvm-test-suite)
Merging now.

**Pull request merged**